src/main/java/com/clouway/asynctaskscheduler/gae/RoutingEventDispatcher.java

```diff
@@ -24,86 +23,176 @@
+    if (validParams(eventClassAsString, eventAsJson, evenHandlerClassName)) {
+
+      AsyncEvent<AsyncEventHandler> event = getAsyncEvent(eventAsJson, eventClassAsString);
 
-    ByteArrayInputStream inputStream = null;
+      AsyncEventHandler handler = handlerFactory.create(event.getAssociatedHandlerClass());
 
-    try {
-      inputStream = new ByteArrayInputStream(eventAsJson.getBytes("UTF-8"));
-    } catch (UnsupportedEncodingException e) {
-      e.printStackTrace();
+      event.dispatch(handler);
     }
+  }
 
-    AsyncEvent<AsyncEventHandler> event = (AsyncEvent) eventTransport.in(eventClass, inputStream);
+  /**
+   * Returns valid async event object if valid parameters are provided
+   * @param eventAsJson
+   * @param eventClassAsString
+   * @return
+   * @throws ClassNotFoundException
+   */
+  private AsyncEvent<AsyncEventHandler> getAsyncEvent(String eventAsJson, String eventClassAsString) throws ClassNotFoundException {
+    Class<?> eventClass = Class.forName(eventClassAsString);
 
-    try {
+    AsyncEvent<AsyncEventHandler> event = null;
 
-      if (inputStream != null) {
-        inputStream.close();
+    if (validEvent(eventClass)) {
+      ByteArrayInputStream inputStream = null;
+
+      try {
+        inputStream = new ByteArrayInputStream(eventAsJson.getBytes("UTF-8"));
+      } catch (UnsupportedEncodingException e) {
+        e.printStackTrace();
+      }
+
+      event = (AsyncEvent) eventTransport.in(eventClass, inputStream);
+
+      try {
+
+        if (inputStream != null) {
+          inputStream.close();
+        }
+      } catch (IOException e) {
+        e.printStackTrace();
       }
-    } catch (IOException e) {
-      e.printStackTrace();
     }
 
     return event;
   }
 
-  private void dispatchListeners(AsyncEvent<AsyncEventHandler> event) {
-    List<? extends AsyncEventListener> listeners  = listenersFactory.create(event.getClass());
-    for (AsyncEventListener listener : listeners) {
-      listener.onEvent(event);
-    }
-  }
 
   /**
-   * Dsipatches the event to it's handler
+   * Dispatches the given event handler and the listeners in separate task queue in datastore transaction if there are no listeners for the event
+   * else dispatches the handler in the same task queue
    * @param event
    * @param evenHandlerClass
    */
-  private void dispatchHandler(AsyncEvent<AsyncEventHandler> event, Class<? extends AsyncEventHandler> evenHandlerClass) {
-    AsyncEventHandler handler = handlerFactory.create(evenHandlerClass);
+  private void dispatchHandlerAndListeners(AsyncEvent<AsyncEventHandler> event, Class<? extends AsyncEventHandler> evenHandlerClass) {
+
+    DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
+
+    Transaction txn = null;
+    try {
+      txn = ds.beginTransaction();
 
-    event.dispatch(handler);
+      List<? extends AsyncEventListener> listeners  = listenersFactory.create(event.getClass());
```

Comment:
we do not need instances here, if we later just use the instance like that - listener.getClass(); 
here we should return the classes of the listeners.